panic nil pointer on file store

Hi, we encountered a panic while running nats-streaming in clustered mode with 3 nodes. 2 are now failing.

```
[1] 2018/07/05 09:24:37.394899 [INF] STREAM: Starting nats-streaming-server[test-cluster] version 0.10.2
[1] 2018/07/05 09:24:37.394940 [INF] STREAM: ServerID: 9B9N8KWFN2AxjijLqN0TYG
[1] 2018/07/05 09:24:37.394944 [INF] STREAM: Go version: go1.10.3
[1] 2018/07/05 09:24:37.402666 [INF] Starting nats-server version 1.1.0
[1] 2018/07/05 09:24:37.402714 [INF] Git commit [not set]
[1] 2018/07/05 09:24:37.417410 [INF] Starting http monitor on 0.0.0.0:8222
[1] 2018/07/05 09:24:37.417506 [INF] Listening for client connections on 0.0.0.0:4222
[1] 2018/07/05 09:24:37.417512 [INF] Server is ready
[1] 2018/07/05 09:24:37.417799 [INF] Listening for route connections on 0.0.0.0:6222
[1] 2018/07/05 09:24:37.418823 [INF] 10.247.73.118:6222 - rid:1 - Route connection created
[1] 2018/07/05 09:24:37.553514 [INF] STREAM: Recovering the state...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x8247cb]

goroutine 64 [running]:
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).ensureLastMsgAndIndexMatch(0xc420084b60, 0xc420073310, 0x0, 0x0, 0x0, 0xb1bd00)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2407 +0xeb
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).recoverOneMsgFile(0xc420084b60, 0xc420073310, 0x1, 0x401, 0x0, 0xc42005dd40)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2260 +0x10bc
github.com/nats-io/nats-streaming-server/stores.(*FileStore).newFileMsgStore(0xc4200b0840, 0xc420022420, 0x16, 0xc420022586, 0x10, 0xc4201715c8, 0x1, 0x153e6ebc2626a4db, 0x153e6ebc2626a4db, 0x0)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1992 +0x8cb
github.com/nats-io/nats-streaming-server/stores.(*FileStore).recoverOneChannel(0xc4200b0840, 0xc420022420, 0x16, 0xc420022586, 0x10, 0xc4201715c8, 0xc42000a120)
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1392 +0xd0
created by github.com/nats-io/nats-streaming-server/stores.(*FileStore).Recover
	/go/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1325 +0x754
```

--- a/server/server.go
+++ b/server/server.go
@@ -47,7 +47,7 @@ import (
 // Server defaults.
 const (
 	// VERSION is the current version for the NATS Streaming server.
-	VERSION = "0.10.2"
+	VERSION = "0.10.3"
 
 	DefaultClusterID      = "test-cluster"
 	DefaultDiscoverPrefix = "_STAN.discover"

--- a/stores/filestore.go
+++ b/stores/filestore.go
@@ -2257,21 +2257,26 @@ func (ms *FileMsgStore) recoverOneMsgFile(fslice *fileSlice, fseq int, useIdxFil
 			offset += msgIndexRecSize
 		}
 		if err == nil {
-			err = ms.ensureLastMsgAndIndexMatch(fslice, lastSeq, lastIndex)
-			if err != nil {
-				ms.fstore.log.Errorf(err.Error())
-				if _, serr := fslice.file.handle.Seek(4, io.SeekStart); serr != nil {
-					panic(fmt.Errorf("File %q: unable to set position to beginning of file: %v", fslice.file.name, serr))
+			if lastIndex != nil {
+				err = ms.ensureLastMsgAndIndexMatch(fslice, lastSeq, lastIndex)
+				if err != nil {
+					ms.fstore.log.Errorf(err.Error())
+					if _, serr := fslice.file.handle.Seek(4, io.SeekStart); serr != nil {
+						panic(fmt.Errorf("File %q: unable to set position to beginning of file: %v", fslice.file.name, serr))
+					}
 				}
+			} else {
+				// Nothing recovered from the index file, try to recover
+				// from data file in case it is not empty.
+				useIdxFile = false
 			}
-		} else {
-			ms.fstore.log.Errorf("Error with index file %q: %v. Truncating and recovering from data file", fslice.idxFile.name, err)
 		}
 		// We can get an error either because the index file was corrupted,
 		// or because the data file is. In both case, we truncate the index
 		// file and recover from data file. The handling of unexpected EOF
 		// is handled in the data file recovery down below.
 		if err != nil {
+			ms.fstore.log.Errorf("Error with index file %q: %v. Truncating and recovering from data file", fslice.idxFile.name, err)
 			if terr := ms.fm.truncateFile(fslice.idxFile, 4); terr != nil {
 				panic(fmt.Errorf("Error during recovery of file %q: %v, you need "+
 					"to manually remove index file %q (truncate failed with err: %v)",
@@ -2283,7 +2288,6 @@ func (ms *FileMsgStore) recoverOneMsgFile(fslice *fileSlice, fseq int, useIdxFil
 			fslice.msgsSize = 0
 			fslice.firstWrite = 0
 			file = fslice.file
-			br = bufio.NewReaderSize(file.handle, defaultBufSize)
 			err = nil
 			useIdxFile = false
 		}
@@ -2293,6 +2297,9 @@ func (ms *FileMsgStore) recoverOneMsgFile(fslice *fileSlice, fseq int, useIdxFil
 		// Get these from the file store object
 		crcTable := ms.fstore.crcTable
 		doCRC := ms.fstore.opts.DoCRC
+
+		// Create a buffered reader from the data file to speed-up recovery
+		br := bufio.NewReaderSize(fslice.file.handle, defaultBufSize)
 
 		// We are going to write the index file while recovering the data file
 		bw := bufio.NewWriterSize(fslice.idxFile.handle, msgIndexRecSize*1000)

--- a/stores/filestore_msg_test.go
+++ b/stores/filestore_msg_test.go
@@ -1693,3 +1693,53 @@ func TestFSExpirationWithTruncatedNonLastSlice(t *testing.T) {
 		t.Fatalf("Expected no message, got %v/%v", n, b)
 	}
 }
+
+func TestFSRecoverEmptyIndexMsgFile(t *testing.T) {
+	cleanupFSDatastore(t)
+	defer cleanupFSDatastore(t)
+
+	s := createDefaultFileStore(t)
+	defer s.Close()
+
+	c := storeCreateChannel(t, s, "foo")
+	ms := c.Msgs
+	storeMsg(t, c, "foo", 1, []byte("msg"))
+	ms.(*FileMsgStore).RLock()
+	datname := ms.(*FileMsgStore).writeSlice.file.name
+	idxname := ms.(*FileMsgStore).writeSlice.idxFile.name
+	ms.(*FileMsgStore).RUnlock()
+	s.Close()
+
+	// Truncate both files to just file header size
+	if err := os.Truncate(datname, 4); err != nil {
+		t.Fatalf("Error truncating %s: %v", datname, err)
+	}
+	if err := os.Truncate(idxname, 4); err != nil {
+		t.Fatalf("Error truncating %s: %v", datname, err)
+	}
+
+	s, rs := openDefaultFileStore(t)
+	defer s.Close()
+
+	c = getRecoveredChannel(t, rs, "foo")
+	if n, b := msgStoreState(t, c.Msgs); n != 0 || b != 0 {
+		t.Fatalf("Unexpected state: %v - %v", n, b)
+	}
+
+	// Files were empty, store one messages
+	msg := storeMsg(t, c, "foo", 1, []byte("msg"))
+	s.Close()
+
+	// Truncate only index file
+	if err := os.Truncate(idxname, 4); err != nil {
+		t.Fatalf("Error truncating %s: %v", datname, err)
+	}
+	// Recover and check that message is recovered (index reconstructed)
+	s, rs = openDefaultFileStore(t)
+	defer s.Close()
+	c = getRecoveredChannel(t, rs, "foo")
+	rm := msgStoreLookup(t, c.Msgs, msg.Sequence)
+	if !reflect.DeepEqual(rm, msg) {
+		t.Fatalf("Expected %v, got %v", msg, rm)
+	}
+}

